Update Openvino API to 2.0

rahulchaphalkar opened the pull request:
Updated to the new API 2.0

Some notes about the tests -
`classify-mobilenet`, `classify-alexnet`, and `classify-inception` required image tensor layout to be `NHWC`.
`detect-inception` required slight modifications in the expected results noted in the comments, like the actual results starting from 1st element instead of 0th.

@abrown

abrown commented on crates/xtask/src/codegen.rs:
```diff
@@ -99,7 +99,9 @@ impl CodegenCommand {
         bindgen::Builder::default()
             .header(header_file.as_ref().to_string_lossy())
-            .allowlist_type("ie_.*")
+            .clang_arg("-I/home/rahul/repos/openvino-binding/openvino-rs/crates/openvino-sys/upstream/src/bindings/c/include")
```
This needs to be a relative path, maybe `-Iopenvino-sys/upstream/src/bindings/c/include`

rahulchaphalkar replied:
done, changed to `.clang_arg("-I./crates/openvino-sys/upstream/src/bindings/c/include")`

abrown commented on crates/xtask/src/codegen.rs:
```diff
@@ -99,7 +99,9 @@ impl CodegenCommand {
             .header(header_file.as_ref().to_string_lossy())
-            .allowlist_type("ie_.*")
+            .clang_arg("-I/home/rahul/repos/openvino-binding/openvino-rs/crates/openvino-sys/upstream/src/bindings/c/include")
+            // .allowlist_type("ie_.*")
```
```suggestion
```

rahulchaphalkar replied:
Done

abrown commented on crates/xtask/src/codegen.rs:
```diff
@@ -132,7 +134,9 @@ impl CodegenCommand {
         bindgen::Builder::default()
             .header(header_file.as_ref().to_string_lossy())
-            .allowlist_function("ie_.*")
+            .clang_arg("-I/home/rahul/repos/openvino-binding/openvino-rs/crates/openvino-sys/upstream/src/bindings/c/include")
```
Ditto

rahulchaphalkar replied:
Done, changed to `.clang_arg("-I./crates/openvino-sys/upstream/src/bindings/c/include")`

abrown commented on crates/xtask/src/codegen.rs:
```diff
@@ -132,7 +134,9 @@ impl CodegenCommand {
             .header(header_file.as_ref().to_string_lossy())
-            .allowlist_function("ie_.*")
+            .clang_arg("-I/home/rahul/repos/openvino-binding/openvino-rs/crates/openvino-sys/upstream/src/bindings/c/include")
+            //.allowlist_function("ie_.*")
```
```suggestion
```

rahulchaphalkar replied:
Done

abrown commented on crates/xtask/src/codegen.rs:
```diff
@@ -156,4 +160,6 @@ impl CodegenCommand {
 const FUNCTIONS_FILE: &str = "functions.rs";
 const DEFAULT_OUTPUT_DIRECTORY: &str = "openvino-sys/src/generated";
-const DEFAULT_HEADER_FILE: &str = "openvino-sys/upstream/src/bindings/c/include/c_api/ie_c_api.h";
+//const DEFAULT_HEADER_FILE: &str = "openvino-sys/upstream/src/bindings/c/include/c_api/ie_c_api.h";
```
```suggestion
```

rahulchaphalkar replied:
Done

abrown commented on crates/openvino/tests/detect-inception.rs:
```diff
@@ -56,14 +92,15 @@ fn detect_inception() {
         .collect();
     results.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
 
+    //Result buffer seem off by 1
```
What is in `results[0]`?

rahulchaphalkar replied:
Removed test as it was outdated - https://github.com/intel/openvino-rs/issues/93

abrown reviewed (commented):
I think this is almost there &mdash; thanks for all this work! Most of my comments are minor nits and easily fixable except for two bigger problems:
- our handling of the `instance: *mut ...` fields needs to be _very safe_: there's a dangling null pointer in there somewhere and I would prefer if we just made all of those fields private to avoid any munging with the pointer. For the structs that are built somewhere else, we can always add a private `fn new(instance: *mut ...) -> Self { Self { instance } }`.
- the pre- and post-processing is just... confusing: maybe just do the best you can with doc tests and comments and we can refactor this later. I see that the C++ has what is called a [`PrePostProcessor`](https://docs.openvino.ai/2023.3/api/c_cpp_api/classov_1_1preprocess_1_1_pre_post_processor.html) which operates on a `Model` and then eventually one `build()`s the new thing. The C API is a bit more confusing so that might be what is going on here.

BTOdell commented:
Really nice to see this work getting done. Thank you!

I'm in the process of porting my company's Python app to Rust, and these API 2.0 bindings are required. I was about to start this conversion process myself since I knew from a couple months ago that these Rust bindings were outdated for the 2024.0 release of OpenVINO.

I've been reviewing this PR and I have some feedback:

1. "Getter" functions should not be prefixed with "get_". This is a standard Rust naming convention: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter
2. There are a couple instances where the struct name is included in the struct's function name and results in overly verbose naming.
3. Functions that take file paths as inputs should use the `&Path` instead of `&str`.

I will continue reviewing this PR and submit any more feedback I have.

Thanks again! 🎉

abrown commented:
@BTOdell, I think I'll take a pass or two after this merges to move this towards "ergonomic Rust." I'll definitely keep your suggestions in mind as I do that (and you're welcome to contribute PRs!). With this PR, I'm more concerned that the basic functionality and correctness remains in place; we can always smooth things out in follow-on PRs before the next release.

BTOdell commented on crates/openvino-sys/upstream:
https://github.com/openvinotoolkit/openvino/releases/tag/2023.3.0

rahulchaphalkar commented on crates/openvino-sys/src/lib.rs:
````diff
@@ -9,12 +9,6 @@
-//! openvino_sys::library::load().expect("to have an OpenVINO library available");
-//! let version = unsafe { CStr::from_ptr(openvino_sys::ie_c_api_version().api_version) };
-//! assert!(version.to_string_lossy().starts_with("2"));
-//! ```
````
Added

abrown replied:
I still don't see this as added back in in the final PR state.

rahulchaphalkar commented on crates/openvino/src/core.rs:
```diff
@@ -1,34 +1,35 @@
+/// See [Core](https://docs.openvino.ai/2023.3/api/c_cpp_api/group__ov__core__c__api.html).
 pub struct Core {
-    instance: *mut ie_core_t,
+    pub(crate) instance: *mut ov_core_t,
```
Made struct fields private, added needed functions

rahulchaphalkar commented on crates/openvino/src/core.rs:
```diff
@@ -1,34 +1,35 @@
+        openvino_sys::library::load().map_err(LoadingError::SystemFailure)?;
         let file = if let Some(file) = xml_config_file {
             cstr!(file.to_string())
         } else if let Some(file) = openvino_finder::find_plugins_xml() {
```
Removed code to find xml file

rahulchaphalkar commented on crates/openvino/src/shape.rs:
```diff
@@ -0,0 +1,60 @@
+
+/// Represents a shape in OpenVINO.
+pub struct Shape {
+    pub(crate) instance: ov_shape_t,
```
Changed fields to private

rahulchaphalkar commented on crates/openvino/tests/classify-alexnet.rs:
```diff
@@ -4,41 +4,89 @@ mod fixtures;
 use util::{Prediction, Predictions};
 
 #[test]
 fn classify_alexnet() {
```
Good idea, changed

rahulchaphalkar commented on crates/openvino/src/prepostprocess.rs:
```diff
@@ -0,0 +1,318 @@
+
+/// The PrePostprocess struct represents pre and post-processing capabilities
+#[derive(Debug)]
+pub struct PrePostprocess {
```
Updated names from `PrePostprocess*` to `PrePostProcess*`

rahulchaphalkar commented on crates/openvino/src/tensor.rs:
```diff
@@ -0,0 +1,216 @@
+/// See [`Tensor`](https://docs.openvino.ai/2023.3/api/c_cpp_api/group__ov__tensor__c__api.html).
+pub struct Tensor {
+    /// Pointer to the underlying OpenVINO tensor.
+    pub instance: *mut ov_tensor_t,
```
Changed fields to private. Where needed, I have added a getter `pub fn instance(&self)-> Result<*mut ov_tensor_t> {Ok(self.instance)}`, as well as function to build from a different module `pub(crate) fn new_from_instance(instance: *mut ov_tensor_t) -> Result<Self> {Ok(Self { instance })}`

rahulchaphalkar commented on crates/openvino/src/element_type.rs:
```diff
@@ -0,0 +1,55 @@
+    fn check_discriminant_values() {
+        assert_eq!(ov_element_type_e_UNDEFINED, ElementType::Undefined as u32);
+        assert_eq!(ov_element_type_e_U8, ElementType::U8 as u32);
+    }
```
Added

rahulchaphalkar commented on crates/openvino/src/prepostprocess.rs:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    /// Builds model with all steps from pre/postprocessing.
+    pub fn build(&self, new_model: &mut Model) -> Result<()> {
```
From my understanding, it doesn't need owned `Model`, it simply builds a new model based on all the preprocessor steps.
`ov_preprocess_prepostprocessor_build(const ov_preprocess_prepostprocessor_t *preprocess, ov_model_t **model)`
I've changed the `build` function and the tests which use it to make it so that it now returns a `new_model`
Here's[ C version](https://github.com/openvinotoolkit/openvino/blob/a7c7de4c65cdb683451d59bf207e93c444030d0d/samples/c/hello_classification/main.c#L202) of `hello_classification` where it does `preprocessor_build`

rahulchaphalkar commented on crates/openvino/src/prepostprocess.rs:
```diff
@@ -0,0 +1,318 @@
+            instance: preprocess_steps,
+        })
+    }
+}
```
I agree, PreProcess operations in general and their usage seem ambiguous here. I've opened a [new issue](https://github.com/intel/openvino-rs/issues/94) which will help track improved docs/examples related to preprocessing.

rahulchaphalkar commented on crates/openvino/tests/detect-inception.rs:
```diff
@@ -1,52 +1,88 @@
+    let input_shape = Shape::new(&vec![1, 481, 640, 3]).unwrap();
+    let element_type = ElementType::U8;
+    let tensor = Tensor::new_from_host_ptr(element_type, input_shape, &data).unwrap();
+    let pre_post_process = PrePostprocess::new(&model).unwrap();
```
These are indeed present in the C-example (not for this particular test, as it is outdated, but for other tests which do have a C/C++ equivalent).
https://github.com/intel/openvino-rs/issues/94 tracks progress on docs/examples for preprocessing

rahulchaphalkar commented:
@abrown Thanks for the detailed review and suggestions, I have addressed your feedback (kept a few conversations `unresolved` in case you wanted to go through my comments). 
I have also opened 2 new issues - https://github.com/intel/openvino-rs/issues/93 to separately keep track on adding a new test similar to `detect-inception`, since `detect-inception` is outdated and based on a demo which no longer exists. 
https://github.com/intel/openvino-rs/issues/94 for adding examples and docs specifically for PreProcessing functionality

@BTOdell thanks for reviewing the PR as well as providing good feedback! I have created a separate issue https://github.com/intel/openvino-rs/issues/95 to keep track of the changes that you suggested in your comment above.

abrown commented on crates/openvino/src/model.rs:
```diff
@@ -16,14 +16,24 @@ use openvino_sys::{
+    }
+
+    /// Get the pointer to the underlying [`ov_model_t`].
+    pub fn instance(&self) -> Result<*mut ov_model_t> {
```
This should be `pub(crate)` to avoid exposing the underlying C types.

abrown commented on crates/openvino/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ exclude = ["/tests"]
 openvino-sys = { path = "../openvino-sys", version = "0.6.0" }
 openvino-finder = { path = "../openvino-finder", version = "0.6.0" }
 thiserror = "1.0"
+anyhow = "1.0"
```
I'm not sure I understand why we would need to import this: just using the regular `Result` with the appropriate error type from this crate should be good enough.

abrown replied:
Oh, I see, you're just using `anyhow` for easier testing... in that case I'll just move it to `[dev-dependencies]` so that it is available during testing but not required by users of the crate.

abrown commented on crates/openvino/src/layout.rs:
```diff
@@ -0,0 +1,44 @@
+
+impl Layout {
+    /// Get [`ov_layout_t`] instance.
+    pub fn instance(&self) -> Result<*mut ov_layout_t> {
```
This should be `pub(crate)` to avoid leaking the underlying C types:

```suggestion
    pub(crate) fn instance(&self) -> Result<*mut ov_layout_t> {
```

abrown commented on crates/openvino/src/lib.rs:
```diff
@@ -53,10 +59,17 @@ pub use tensor_desc::TensorDesc;
-        .into_owned();
-    unsafe { openvino_sys::ie_version_free(std::ptr::addr_of_mut!(ie_version)) };
-    str_version
+    //let mut ov_version = std::ptr::null_mut();
```
```suggestion
```

abrown commented on crates/openvino/src/model.rs:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /// Get the pointer to the underlying [`ov_model_t`].
+    pub fn instance(&self) -> Result<*mut ov_model_t> {
```
```suggestion
    pub(crate) fn instance(&self) -> Result<*mut ov_model_t> {
```

abrown commented on crates/openvino/src/model.rs:
```diff
@@ -0,0 +1,115 @@
+    pub fn instance(&self) -> Result<*mut ov_model_t> {
+        Ok(self.instance)
+    }
+
+    /// Create a new instance of the Model struct.
+    pub fn new() -> Result<Self> {
+        let instance = std::ptr::null_mut();
+        Ok(Self { instance })
+    }
```
It seems highly unsafe to have a model containing a null pointer.

```suggestion
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+
+impl Port {
+    ///Get [`ov_output_const_port_t`] instance.
+    pub fn instance(&self) -> Result<*mut ov_output_const_port_t> {
```
```suggestion
    pub(crate) fn instance(&self) -> Result<*mut ov_output_const_port_t> {
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+}
+
+impl Port {
+    ///Get [`ov_output_const_port_t`] instance.
```
```suggestion
    /// Get [`ov_output_const_port_t`] instance.
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+use openvino_sys::{ov_output_const_port_t, ov_port_get_any_name};
+use std::ffi::CStr;
+
+///Struct for Port of a model.
```
```suggestion
/// Struct for Port of a model.
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+        Ok(self.instance)
+    }
+
+    ///Create a new [`Port`] from [`ov_output_const_port_t`].
```
```suggestion
    /// Create a new [`Port`] from [`ov_output_const_port_t`].
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+    pub(crate) fn new(instance: *mut ov_output_const_port_t) -> Result<Self> {
+        Ok(Self { instance })
+    }
+    ///Get name of a port.
```
```suggestion
    /// Get name of a port.
```

abrown commented on crates/openvino/src/port.rs:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    ///Create a new [`Port`] from [`ov_output_const_port_t`].
+    pub(crate) fn new(instance: *mut ov_output_const_port_t) -> Result<Self> {
```
No need to return a `Result` when there is no chance for an error.
```suggestion
    pub(crate) fn new(instance: *mut ov_output_const_port_t) -> Self {
```

abrown commented on crates/openvino/src/shape.rs:
```diff
@@ -0,0 +1,67 @@
+
+impl Shape {
+    /// Get the pointer to the underlying OpenVINO shape.
+    pub fn instance(&self) -> ov_shape_t {
```
```suggestion
    pub(crate) fn instance(&self) -> ov_shape_t {
```

abrown reviewed (commented):
I think the only thing to resolve is the recent addition of `anyhow`; let's talk offline about that.

abrown commented on crates/openvino/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ exclude = ["/tests"]
 openvino-sys = { path = "../openvino-sys", version = "0.6.0" }
 openvino-finder = { path = "../openvino-finder", version = "0.6.0" }
 thiserror = "1.0"
+anyhow = "1.0"
 
 [dev-dependencies]
 float-cmp = "0.9"
```
```suggestion

[dev-dependencies]
float-cmp = "0.9"
anyhow = "1.0"
```

abrown commented on crates/openvino/src/shape.rs:
```diff
@@ -0,0 +1,67 @@
+        Ok(Self { instance: shape })
+    }
+
+    /// Create a new shape object from ov_shape_t.
+    pub(crate) fn new_from_instance(instance: ov_shape_t) -> Result<Self> {
+        Ok(Self { instance })
+    }
```
```suggestion
    /// Create a new shape object from `ov_shape_t`.
    pub(crate) fn new_from_instance(instance: ov_shape_t) -> Self {
        Self { instance }
    }
```

abrown commented on crates/openvino/src/shape.rs:
```diff
@@ -0,0 +1,67 @@
+        Ok(Self { instance })
+    }
+
+    /// Returns the rank of the shape.
+    pub fn get_rank(&self) -> Result<i64> {
+        Ok(self.instance.rank)
+    }
```
Here and many other places we return a `Result` unnecessarily.
```suggestion
    /// Returns the rank of the shape.
    pub fn get_rank(&self) -> i64 {
        self.instance.rank
    }
```

abrown reviewed (approved):
I fixed up some of the small stuff but can't push to this branch so I can't get rid of `Model::new` and unnecessary returns of `Result`. Otherwise looks great!

abrown reviewed (approved):
Thanks for all the work on this!
